### PR TITLE
Enable or disable image pushing

### PR DESCRIPTION
**Pull request by bartbutenaers**
Hi Riku (@rikukissa),

I currently use your contribution for displaying image data, identical like I use the **Debug** node for displaying non-image data (json, buffers, ...).

But the **Debug** node has a nice extra feature, which is enable or disable displaying data: by simply pressing a button.  I have reused part of their [js](https://github.com/node-red/node-red/blob/master/nodes/core/core/58-debug.js) and [html](https://github.com/node-red/node-red/blob/master/nodes/core/core/58-debug.html) code to add a similar button to your node.

When the button is pressed, it will go half way beneath the node and no more images will be pushed from the server.  When the button is pressed again, it will become entirely visible again and images will again be pushed from the server:

image

I would appreciate if you could test this, and have a look whether you want to add this functionality to your new version.

P.S. not all of the history-related code in the html is clear to me, but it seems to do the job ...

Thanks a lot,
Bart

### CODE REVIEW

**Comment by rikukissa** [comment]
Good idea! I'll go through the code in more detail as soon as I can.

**Comment by rikukissa** [comment]
Quick question: does the active state persist between restarts of node-red?

**Comment by bartbutenaers** [comment]
Quick answer: I added a mix of buttons (enabled and disabled) to my flows and their state was persisted.  Did some redeploys and even after a restart of my Node-Red, all buttons maintained their status.
Enough for today ;-)

File: image/image.js
Comment (rikukissa) [inline image/image.js]: I think in both cases of state, the response status should be 200. 201 is commonly used to indicate a new resource was created, which isn't the case here.  Could we use some other way of distinguishing the proceeding state on the frontend code?

File: image/image.html
Comment (rikukissa) [inline image/image.html]: Use `const` instead of `var`

File: image/image.html
Comment (rikukissa) [inline image/image.html]: Would it make sense to move these response handlers somewhere outside the `onclick` handler scope?

File: image/image.html
Comment (rikukissa) [inline image/image.html]: `const`

File: image/image.js
Comment (rikukissa) [inline image/image.js]: maybe do 

```js
const nodeExists = node !== null && typeof node !== "undefined"

if(!nodeExists) {
  res.sendStatus(404);
  return;  
}

if (state === "enable") {
  node.active = true;
  res.sendStatus(200);
} else if (state === "disable") {
  node.active = false;
  res.sendStatus(201);
} else {
  res.sendStatus(404);
}
```
to reduce the needed indentation

File: image/image.html
Comment (rikukissa) [inline image/image.html]: Would it make sense to abstract the `RED.notify(node._("common.notification.error", {message: ` part to a function to reduce repetition?

File: image/image.html
Comment (rikukissa) [inline image/image.html]: Lets keep the indentation as 2 spaces. At least VSCode's `Format document` will fix these automatically

File: image/image.html
Comment (rikukissa) [inline image/image.html]: Use the same spacing as below

File: image/image.html
Comment (rikukissa) [inline image/image.html]: [Template literal](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals) would probably make this look a bit nicer :)

**Review by rikukissa — CHANGES_REQUESTED** [review]
Few comments, but overall it looks good and does what it promises. Great work 👍

File: image/image.js
Comment (bartbutenaers) [reply image/image.js]: Hi Riku,

A little bit later as expected, but it's a busy week ...
As I mentioned in the beginning, I based this pull request on the Debug node.  Hadn't been reviewing the Debug code in depth, since it is written by the core team.

But indeed you are right.  In the latest version, I just use send('deactived') to let the flow editor know that the server side node has been deactived.  In that case the status is automatically set to 200:

image

This makes the html also simple: I just show the "deactivated" (or "activated") immediately in the popup notification.

File: image/image.html
Comment (bartbutenaers) [reply image/image.html]: Done

File: image/image.html
Comment (bartbutenaers) [reply image/image.html]: Haven't done that.  Not really sure what you want to achieve with it.  Readability ??

File: image/image.html
Comment (bartbutenaers) [reply image/image.html]: Done

File: image/image.html
Comment (bartbutenaers) [reply image/image.html]: Done.

File: image/image.html
Comment (bartbutenaers) [reply image/image.html]: I agree ...

File: image/image.html
Comment (bartbutenaers) [reply image/image.html]: Start using VSCode is on my TODO list.  But the list is growing ;-)

File: image/image.html
Comment (bartbutenaers) [reply image/image.html]: Yep, I have changed it.  I have also replaced the locale paths (like e.g.  *common.notification.error*) by a fixed text.  Because the locales might change in the future, resulting in corrupting our code.   Have used your template literals all over the place ;-)

Here are some test results:

image

File: image/image.js
Comment (bartbutenaers) [reply image/image.js]: Makes sense for readability.  As mentioned below, I have changed the sendStatus for 200 en 201 by send (which automatically sets the status to 200).

**Comment by bartbutenaers** [comment]
@rikukissa ,

Sorry for the delay, but it is rather busy at the moment ...

The only thing I find confusing about this button: when your node is located on a **disabled** flow tab, the user will get a popup **"node is not deployed"** when he presses the button.  This is in fact correct: when the flow tab is disabled, the nodes of that flow are not deployed (so the nodeexists == false).  So you cannot (de)activate such a node with the button.

But the users will not always immediately think about the flow tab being disabled.  
Any suggestions on this one?? Some possible solutions:
+ Leave it this way
+ Try to find out if the flow is disabled, and in that case show another error "Flow is disabled"
+ Show always more information in the error "Node is not deployed or flow is disabled"
+ Disable the button when the tab is disabled.  But that is again confusing, because the user doesn't know why the button is disabled.
+ ...

But the Debug node has the same behaviour!!!   

Bart

**Comment by bartbutenaers** [comment]
Hi Riku (@rikukissa ),

I'm going to wait with my 2 new pull requests (one for displaying image properties, and one for the grab handle) until you have approved and merged this one into your master.  Otherwise my fork is going to become a mess...

Hope to hear from you soon,
Bart

**Comment by bartbutenaers** [comment]
Hello Riku (@rikukissa ),

Did you had already a chance to have a look at the changes.  Would be great if we could finish the pull requests, so you can get rid of your poltergeist ;-)

Thanks,
Bart

**Pull request closed without merging**
